trace/execution_state.go

```diff
@@ -45,14 +74,16 @@ type WorkflowExecutionState struct {
 
 	// ChildStates contains all the ExecutionStates contained by this WorkflowExecutionState in order of execution.
 	ChildStates []ExecutionState
-	// activityMap contains all the activities executed in the Workflow, indexed by the EVENT_TYPE_ACTIVITY_TASK_SCHEDULED event id.
-	// Used to retrieve the activities from events.
+	// childMapLock locks the state for read/write access. This is mainly used to avoid concurrent read/writes on the child maps (activityMap, childWorkflowMap and timerMap).
+	childMapLock sync.RWMutex
```

Comment:
Are these the only properties that will be accessed concurrently?
If you're certain of this, let's keep the way it is but otherwise, I'd put a lock on the entire struct.

Reply:
There's no problem with properties being accessed concurrently, only maps.

Reply:
It depends if we care about multiple properties being updated atomically.
I generally would like to make less assumptions and be on the safe side.